runtime/common/src/lib.rs

```diff
@@ -242,6 +228,67 @@ pub mod parachains {
 	}
 }
 
+/// CurrencyIdConvert
+/// This type implements conversions from our `CurrencyId` type into `MultiLocation` and vice-versa.
+/// A currency locally is identified with a `CurrencyId` variant but in the network it is identified
+/// in the form of a `MultiLocation`, in this case a pCfg (Para-Id, Currency-Id).
+pub struct CurrencyIdConvert<ParachainId, AssetRegistry>(sp_std::marker::PhantomData<(ParachainId, AssetRegistry)>);
+
+impl<ParachainId: Get<ParaId>, AssetRegistryInspector: AssetRegistryInspect> Convert<CurrencyId, Option<MultiLocation>> for CurrencyIdConvert<ParachainId, AssetRegistryInspector> {
+	fn convert(id: CurrencyId) -> Option<MultiLocation> {
+		AssetRegistryInspector::metadata(&id)
+			.filter(|m| m.location.is_some())
+			.and_then(|m| m.location)
+			.and_then(|l| l.try_into().ok())
+	}
+}
+
+impl<ParachainId: Get<ParaId>, AssetRegistryInspector: AssetRegistryInspect> Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert<ParachainId, AssetRegistryInspector> {
+	fn convert(location: MultiLocation) -> Option<CurrencyId>  {
+		let para_id = ParachainId::get();
+
+		let unanchored_location = match location {
+            MultiLocation { parents: 0, interior } => {
+     
+                match interior.pushed_front_with(Parachain(u32::from(para_id))) {
+                    Ok(new_interior) => MultiLocation {
+                        parents: 1,
+                        interior: new_interior,
+                    },
+                    Err(_) => return None, 
+                }
+            },
+            x => x,
+        };
+		AssetRegistryInspector::location_to_asset_id(unanchored_location)
+	}
+}
+
+impl<ParachainId: Get<ParaId>, AssetRegistryInspector: AssetRegistryInspect> Convert<MultiAsset, Option<CurrencyId>> for CurrencyIdConvert<ParachainId, AssetRegistryInspector> {
+	fn convert(a: MultiAsset) -> Option<CurrencyId> {
+		if let MultiAsset { id: AssetId::Concrete(id), fun: _ } = a {
+			<Self as Convert<MultiLocation, Option<CurrencyId>>>::convert(id)
+		} else {
+			None
+		}
+	}
+}
+
+/// Convert an incoming `MultiLocation` into a `CurrencyId` if possible.
+/// Here we need to know the canonical representation of all the tokens we handle in order to
+/// correctly convert their `MultiLocation` representation into our internal `CurrencyId` type.
+impl<ParachainId: Get<ParaId>, AssetRegistryInspector: AssetRegistryInspect> xcm_executor::traits::Convert<MultiLocation, CurrencyId> for CurrencyIdConvert<ParachainId, AssetRegistryInspector> {
+	fn convert(location: MultiLocation) -> Result<CurrencyId, MultiLocation> {
+		<CurrencyIdConvert<ParachainId, AssetRegistryInspector> as Convert<MultiLocation, Option<CurrencyId>>>::convert(location)
+			.ok_or(location)
+	}
+}
+
+pub trait AssetRegistryInspect {
```

Comment:
I don't think that we need to define this trait and then implement it in the `xcm_config.rs` files of the three runtimes. Instead we can just refer to the asset registry of each runtime directly by using the trait 

```
orml_traits::Inspect<
  AssetId = CurrencyId,
  Balance = Balance,
  CustomMetadata = CustomMetadata,
>
```

Then use the trait methods `asset_id` and `metadata`. The asset registry implements the trait [here](https://github.com/open-web3-stack/open-runtime-module-library/blob/4ae0372e2c624e6acc98305564b9d395f70814c0/asset-registry/src/impls.rs#L176-L196).

Then in each runtime you would just use instead of  `CurrencyIdConvert<AssetRegistry>` instead of `CurrencyIdConvert<ParachainId, AssetRegistryInspector>` (see my other comment why `ParachainId` is not required in my opinion).

Reply:
Thanks for pointing into the trait!, I thought none was implemented to fetch the metadata. I will remove the unnecessary definition of our trait.